src/portable/espressif/esp32s2/dcd_esp32s2.c

```diff
@@ -0,0 +1,751 @@
+  //   * 2 locations for OUT endpoint control words.
+  //   * 16 for largest packet size of 64 bytes. ( TODO Highspeed is 512 bytes)
+  //   * 1 location for global NAK (not required/used here).
+  //   * It is recommended to allocate 2 times the largest packet size, therefore
+  //   Recommended value = 10 + 1 + 2 x (16+2) = 47 --> Let's make it 52
+  USB0.grstctl |= 0x10 << USB_TXFNUM_S; // fifo 0x10,
+  USB0.grstctl |= USB_TXFFLSH_M;        // Flush fifo
+  USB0.grxfsiz = 52;
+
+  // Control IN uses FIFO 0 with 64 bytes ( 16 32-bit word )
+  USB0.gnptxfsiz = (16 << USB_NPTXFDEP_S) | (USB0.grxfsiz & 0x0000ffffUL);
+
+  // Ready to receive SETUP packet
+  USB0.out_ep_reg[0].doeptsiz |= USB_SUPCNT0_M;
+
+  USB0.gintmsk |= USB_IEPINTMSK_M | USB_OEPINTMSK_M;
+}
+
+static void enum_done_processing(void)
+{
+  ESP_EARLY_LOGV(TAG, "dcd_int_handler - Speed enumeration done! Sending DCD_EVENT_BUS_RESET then");
+  // On current silicon on the Full Speed core, speed is fixed to Full Speed.
+  // However, keep for debugging and in case Low Speed is ever supported.
+  uint32_t enum_spd = (USB0.dsts >> USB_ENUMSPD_S) & (USB_ENUMSPD_V);
+
+  // Maximum packet size for EP 0 is set for both directions by writing DIEPCTL
+  if (enum_spd == 0x03) { // Full-Speed (PHY on 48 MHz)
+    USB0.in_ep_reg[0].diepctl &= ~USB_D_MPS0_V;    // 64 bytes
+    USB0.in_ep_reg[0].diepctl &= ~USB_D_STALL0_M; // clear Stall
+    xfer_status[0][TUSB_DIR_OUT].max_size = 64;
+    xfer_status[0][TUSB_DIR_IN].max_size = 64;
+  } else {
+    USB0.in_ep_reg[0].diepctl |= USB_D_MPS0_V;     // 8 bytes
+    USB0.in_ep_reg[0].diepctl &= ~USB_D_STALL0_M; // clear Stall
+    xfer_status[0][TUSB_DIR_OUT].max_size = 8;
+    xfer_status[0][TUSB_DIR_IN].max_size = 8;
+  }
+}
+
+
+/*------------------------------------------------------------------*/
+/* Controller API
+ *------------------------------------------------------------------*/
+void dcd_init(uint8_t rhport)
+{
+  (void)rhport;
+
+  ESP_LOGV(TAG, "DCD init - Start");
+
+  // A. Disconnect
+  ESP_LOGV(TAG, "DCD init - Soft DISCONNECT and Setting up");
+  USB0.dctl |= USB_SFTDISCON_M; // Soft disconnect
+
+  // B. Programming DCFG
+  /* If USB host misbehaves during status portion of control xfer
+    (non zero-length packet), send STALL back and discard. Full speed. */
+  USB0.dcfg |= USB_NZSTSOUTHSHK_M | // NonZero .... STALL
+      (3 << 0);            // dev speed: fullspeed 1.1 on 48 mhz  // TODO no value in usb_reg.h (IDF-1476)
+
+  USB0.gahbcfg |= USB_NPTXFEMPLVL_M | USB_GLBLLNTRMSK_M; // Global interruptions ON
+  USB0.gusbcfg |= USB_FORCEDEVMODE_M;                    // force devmode
+
+  USB0.gotgctl &= ~(USB_BVALIDOVVAL_M | USB_BVALIDOVEN_M | USB_VBVALIDOVVAL_M); //no overrides
+#ifdef CONFIG_IDF_TARGET_ESP32S2BETA // needed for beta chip only
+  //C. chip 7.2.2 hack
+  ESP_LOGV(TAG, "DCD init - chip ESP32-S2 beta hack");
+  USB0.gotgctl = (0 << USB_BVALIDOVVAL_S); //B override value
+  ets_delay_us(20);
```

Comment:
@igrr since removing delay at the end of this function work. Maybe remove this couple of delay_us(20) also work as well. However, there is a any comment on this particular code with BETA mcu. Since I don't have a beta mcu (I am not sure if I have one). Would you mind confirming whether we could safely remove this with Beta, or Beta requires this to work.

Reply:
From what I remember, beta needed this as there was a bug in the phy logic that made the chip reset as soon as it transmitted anything. Setting the B override value made it ignore resets (any resets, generated by the faulty logic or not), which 'fixed' the problem well enough to test usb with it. For the final silicon, this is not needed.

Reply:
AH thanks, then we leave it here, however, do you know any ID register we can check to know it is a beta mcu and run this walkaround instead of depending on user to define the `CONFIG_IDF_TARGET_ESP32S2BETA` macro. If not then never mind, we can leave it as it is.

Reply:
Within esp-idf, this define is generated (in sdkconfig.h, which is only indirectly included here... could use a direct inclusion tbh) when the user compiles for the ESP32S2. As firmwares for the beta and the final silicon are not interchangable in the first place (iirc, @igrr correct me on that if I'm wrong there) there is no need for runtime figuring-this-out-ness anyway. 

(Also, do note that the beta silicon run was very small and software support for it is not in mainstream esp-idf, as such you may consider phasing out support for it alltogether somewhere in the future.)

Reply:
Ah thanks, got it, let's just leave it as it for now, we could safely remove it later on when all beta users got their hand on production.

Reply:
Can safely remove the code related to the beta chip. In case we need it, can restore from git history.

Reply:
@igrr Ok, let's remove it then :)